### PR TITLE
WIP: Hook logs

### PR DESCRIPTION
**Pull request by johnw188**
I wanted to throw this up before I finish work for #2298 as I'm looking for some feedback on my approach and overall design.

From a tiller standpoint, publishing a streamable log is accomplished by making a call to the `PubLog` function on the log streamer pubsub.

```go
s.logs.PubLog(req.Name, release.LogSource_SYSTEM, release.LogLevel_INFO, "Got release status for the release")
```

When one or many gRPC streams are open and listening for logs, making a call to `PubLog` ships the log object to all of the listeners. If no listeners are configured for that combination of log type and level, calling PubLog does nothing (no caching is taking place in this initial implementation).

Consuming logs from the helm client side is done via

```go
	done := make(chan struct{})
	stream, err := l.client.ReleaseLogs(l.release, release.LogLevel_DEBUG, done, release.LogSource_SYSTEM, release.LogSource_POD)
	if err != nil {
		return prettyError(err)
	}

	fmt.Println("Listening for logs")
	for {
		select {
		case l, ok := <-stream:
			if !ok {
				return nil
			}
			fmt.Println(l.Log.Log)
		}
	}
```

(note to self - `l.Log.Log` needs renaming).

I'm particularly interested in style/idiomatic golang and gRPC suggestions, as I'm new to the language. I'm also interested in better names for some of the pieces - I had a lot of trouble finding good names for the log pubsub functions/types.

### CODE REVIEW

**Comment by johnw188** [comment]
CI is broken because I didn't update the testing mocks with the new methods, will fix that also.

File: pkg/tiller/release_server.go [hunk 1]
```diff
@@ -251,6 +254,13 @@ func (s *ReleaseServer) GetReleaseStatus(c ctx.Context, req *services.GetRelease
 		Info:      rel.Info,
 	}
 
+	logOptions := kube.NewLogOptions()
```
Comment (johnw188) [inline pkg/tiller/release_server.go]: Hardcoded test call, ignore

**Comment by thomastaylor312** [comment]
@johnw188 Do you think this will be ready in time for 2.5? Do you need any help or feedback from people?

**Comment by johnw188** [comment]
@thomastaylor312 I would like to get this done for 2.5. I need a code review from someone, and I need to finish up the kube client lib calls to grab logs

**Comment by thomastaylor312** [comment]
@johnw188 When it is ready for review, remove the WIP and ping me. I'll add this to 2.5 so it doesn't fall through the cracks

**Comment by thomastaylor312** [comment]
@johnw188 Do you have a status update on this?

**Comment by johnw188** [comment]
@thomastaylor312 I am taking another crack at this over the weekend. Commented in #2298, streaming is solid but actually querying the logs from tiller has been super tough.

**Comment by thomastaylor312** [comment]
@johnw188 This is already getting pretty big, so your comment in the issue about splitting into 2 PRs may be a good idea. It will also make it easier for others to help (like me 😄)

**Comment by bacongobbler** [comment]
ping @johnw188, are you willing to split this into separate PRs as per @thomastaylor312's request?

**Comment by johnw188** [comment]
@bacongobbler yes, let me update the branch and make sure the build is working

**Comment by johnw188** [comment]
Apologies for leaving a giant PR hanging, I changed jobs recently and I've been figuring out my new situation. I pulled in all the latest changes and completed the merge, which was surprisingly pretty straightforward. I took a look at what I've got here and it falls into a couple categories:

1. Log streaming - complete pending review. This is the addition of the gRPC code to support streaming logs back to the helm client in real time. As mentioned at the start, there's a `PubLog` function on the log pubsub that can be called like so:

```golang
s.logs.PubLog(req.Name, release.LogSource_SYSTEM, release.LogLevel_INFO, "Logged!")
```

On the client side you stream logs by doing:

```golang
done := make(chan struct{})
stream, err := l.client.ReleaseLogs(l.release, release.LogLevel_DEBUG, done, release.LogSource_SYSTEM, release.LogSource_POD)
if err != nil {
	return prettyError(err)
}

fmt.Println("Listening for logs")
for {
	select {
	case l, ok := <-stream:
		if !ok {
			return nil
		}
		fmt.Println(l.Log.Log)
	}
}
```

The `PubLog` call on the tiller side is a noop if nobody is currently listening to that log source at the appropriate log level. This works in real time as expected and is pretty cool and useful to provide to other engineers even without the rest of the work being complete.

2. Querying kubernetes logs from tiller

This was super painful 6 months ago, to the point where I got really frustrated and gave up. I see they restructured the apimachinery packages from the updates to the glide lockfile, so I'm hoping they made it easier to make the logs call from code. This is the work in the `pkg/kube/pod_logs.go` file, and is probably throwaway at this point - I was trying to mimic kubectl as the `kubectl logs` command does exactly what I need to do.

3. Determining the pods that were created by a chart

I was working under the assumption that there was some tiller function I could call to get the pods that had been created by a chart, but that was incorrect. As of a few months back tiller has access to the manifests it created but not the pods that result from executing those manifests I think this functionality would be a useful helper method for more than just this usecase, but I'm not sure the best approach to take in order to make it work.

I'm a bit nervous about just filtering on the deployment/job metadata as it feels like a poorly written chart could accidentally interact with non chart managed resources if someone screws up their metadata? I'm not sure if that concern is founded.

**Comment by johnw188** [comment]
I'm going to grab some sleep and then split these out tomorrow. Should I do new pull requests or leave the log streaming in this one and open new PRs for the other stuff?

**Comment by thomastaylor312** [comment]
@johnw188 Thanks for the update. We can do either way. If you feel like it would be cleaner to do completely new PRs, feel free to do that (just make sure to ping me on them so I can help review). However, I think leaving the log streaming in this PR would probably be easiest and then you can split out the other features.

As for determining pods, that may be some good logic to abstract out because other functions of tiller could use that.

**Comment by thomastaylor312** [comment]
Bumping to 2.8

**Comment by bacongobbler** [comment]
pulling off the 2.8 milestone

**Comment by siwyd** [comment]
Would sure appreciate this functionality making it into Helm!

**Pull request closed without merging**